driver-async/src/main/com/mongodb/async/client/Observer.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2015 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.async.client;
+
+/**
+ * Provides a mechanism for receiving push-based notifications.
+ *
+ * <p>
+ * Will receive a call to {@link #onSubscribe(Subscription)} on subscription to the Observable.
+ * No further notifications will be received until {@link Subscription#request(long)} is called.
+ * </p>
+ * <p>
+ * After signaling demand:
+ * <ul>
+ *   <li>One or more invocations of {@link #onNext(Object)} up to the maximum number defined by {@link Subscription#request(long)}</li>
+ *   <li>Single invocation of {@link #onError(Throwable)} or {@link Observer#onComplete()} which signals a terminal state after which no
+ * further events will be sent.</li>
+ * </ul>
+ * <p>
+ * Demand can be signaled via {@link Subscription#request(long)} whenever the {@link Observer} instance is capable of handling more.
+ *</p>
+ *
+ * @param <TResult> The type of element signaled.
+ * @since 3.1
+ */
+public interface Observer<TResult> {
+
+    /**
+     * Invoked on subscription to an Observable.
```

Comment:
Its difficult as an Observable is anything that can be observed in our case - MongoIterable or any method that takes a SingleResultCallback.